### ISSUE TITLE
Add a `module-info.java`

### ISSUE THREAD

**Issue by sgammon** [protocolbuffers/protobuf#16133]
**What language does this apply to?**
If it's a proto syntax change, is it for proto2 or proto3? `N/A`
If it's about generated code change, what programming language? `Java`

**Describe the problem you are trying to solve.**
I want to use Protobuf Java with [`jlink`](https://docs.oracle.com/en/java/javase/11/tools/jlink.html). Unfortunately, `jlink` rejects JARs with an `Automatic-Module-Name` set; it is required to define a `module-info.java`.

**Describe the solution you'd like**
I would like Protobuf Java to ship a `module-info.java`, in an MRJAR, to preserve existing JDK support. I am not proposing a change to the JVM bytecode minimum support level.

**Describe alternatives you've considered**
I have considered solutions like Moditect, but they complicate my build greatly because dependencies must be transformed before being consumed by the Java Compiler.

**Additional context**
[Guava is considering](https://github.com/google/guava/issues/2970) [a PR for this](https://github.com/google/guava/issues/2970). Recently, [Error Prone](https://github.com/google/error-prone/releases/tag/v2.26.1) and other core libs have released new versions with a `module-info.java` definition. Thus, Protobuf's transitive dependencies may be ready for a fully modular Java build.

**Comment by sgammon** [comment]
Bump!

**Comment by odinsbane** [comment]
I don't understand why this can't be done. It seems like jlink is a great way to create a stand alone app, but we cannot use protobuf with it.

**Comment by vietj** [comment]
We are willing to help on this effort, we would like very much to have JPMS support for Google Protobuf library

**Comment by sgammon** [comment]
@googleberg / @shaod2 Can this be reopened, at least to track the issue? It is not stale

**Comment by sgammon** [comment]
Thank you @zhangskz

**Comment by vietj** [comment]
@sgammon you should update the README of this issue with the recent Guava adoption of JPMS

**Comment by esrauchg** [comment]
Just to give an official update, we don't currently have any active intent to pursue this at this time, as we are not seeing it get enough adoption in the broader Java ecosystem to warrant support effort, and in the broader Google libraries context it would be very difficult to move from the status quo to a model where only one jar may contain classes of a given package.

We intend to keep an eye on this and revisit as things evolve.

We understand that may be frustrating/unfortunate for the people who have adopted technologies like jlink; I hope it is better that we are transparent rather than just continuing to leave the issue in silent limbo.

**Comment by sgammon** [comment]
@esrauchg That is circular logic, because many applications or libraries cannot fully adopt JPMS until their entire suite of dependencies builds on the modulepath. Protobuf of course is a transitive dependency to many libraries, making it a transitive dependency in many applications.

Protobuf, like Guava before it (and no longer), is effectively _blocking adoption of JPMS_, and so deciding not to support JPMS because of a lack of such adoption is of course circular. We can't all adopt JPMS to show Google we are willing to adopt JPMS if adopting JPMS is made impossible because your app transitively uses Protobuf.

Every Java or Kotlin application I have ever built -- and there are many -- has had Protobuf Java on the classpath. Exactly 0 of those apps can support JPMS, and the only way to fix that is to remove Protobuf entirely. That seems like a suboptimal experience, no?

**Comment by sgammon** [comment]
If Google has admitted that Guava should support JPMS because of strong demand, why not protocol buffers? If the community is willing to do this work, could Google put in the work to review it?

**Comment by esrauchg** [comment]
I think concern here is the problem of additional constraints implied by supporting JMPS rather than the up front technical work of the declaration, including eg the implication of a module owning a package (by design you can't have multiple modules make visible the same package).

That particular aspect be more of a hard blocking concern against adoption of JPMS for other Google published libraries like the Java Google Cloud Client Libraries than it is for Protobuf, but I believe this already breaks down even for how Protobuf currently packages our code for our Maven releases where `com.google.protobuf` package is used by several different artifacts.

**Comment by sgammon** [comment]
@esrauchg Okay, I understand, and that's a fantastic point, but JPMS came out with Java 9 in September 2017. It's 2025 now, meaning 8 years have elapsed, and we are now on Java 24. There are no plans to deprecate JPMS. When is the appropriate time to start _thinking_ about that support if not now?

**Comment by sgammon** [comment]
For many Java developers, Google is effectively abdicating its role as a developer tools and libraries provider by not supporting JPMS. If I have to choose between JPMS and Google's libraries, I may be forced to choose against Google's libraries altogether. This has already happened in several projects for us.

So while I understand and even empathize with the challenge that generating code and supporting JPMS will be, I also know that if this work doesn't start by JDK 24, it probably just won't happen. Protobuf is inherently intertwined into everything Google does. Thus, it's reasonable to conclude that it will never happen, and the choice here really is between modern Java and using Protobuf at all. Am I correct in understanding the path Google is encouraging me to take? I'm just trying to interpret what you've said here over time as it relates to Google's own libraries in Java (i.e. Cloud, Gemini...) and our adoption thereof.

**Comment by esrauchg** [comment]
Thank you for your feedback.

Unfortunately Google does not currently have a centralized decision about this, so you will see some patchwork unevenness on the topic among Google's offerings.

I think at this time we haven't heard the same shape/degree of feedback as this from other customers about the value/importance of supporting JPMS.

We will keep this issue open to continuously gather feedback and be sure to speak to other customers to keep a pulse on this topic.

**Comment by odinsbane** [comment]
I think the focus on jlink is a bit of a red herring. It is the tool that made the issue apparent, but a lot of libraries and tools have migrated to be module compatible.  What if we started from the premise that having a module is better.  Does @esrauchg have a specific example whey they say

> but I believe this already breaks down even for how Protobuf currently packages our code for our Maven releases where com.google.protobuf package is used by several different artifacts.

What would break?

**Comment by sgammon** [comment]
Google deciding whether to support JPMS:

![Image](https://github.com/user-attachments/assets/37276970-ecb2-4569-8df3-f26caa1c4ef9)

**Comment by esrauchg** [comment]
@sgammon Please keep the discussion on the github issues professional and productive please, even when you disagree with local technical choices.

> What would break?

JPMS inherently disallows two jars to export the same package (disallows split packages). Which unfortunately turns a legitimate best-practice into a "mandatory, fundamentally cannot violate" state. As far as I understand this would end up being not just a constraint against the current state, but also a forward constraint on potentially breaking one preexisting thing up into two modules while trying to minimize breaking changes by not changing the packages of the classes/interfaces (and this seems especially unfortunate in Java which doesn't have any form of re-export syntax like Rust or Go does that could be used to take the edge off of such changes by enabling you to move the definitions to a new package while also still continuing to export them in the old package).

**Comment by sgammon** [comment]
@esrauchg It is professional, and it is productive, I just felt an image made my point instead of explaining it in words. If we can't adopt JPMS because of Protobuf, we can't really demonstrate to Google demand for JPMS so they support JPMS.

I know you disagree with my point here, or at least it is not motivating to Google, and it would be cool if Google could avoid resorting to moderating my comments merely because they don't align with JPMS strategy (or lack thereof)

**Comment by sgammon** [comment]
We are a funded AI startup and we want to use Gemini and Cloud. We cannot do that at this time because we are adopters of JPMS, which renders Google's Java cloud libraries unusable for us, since literally all of them rely on Protocol Buffers and merely having these _anywhere transitively in our dependency graph_ means we must choose between JPMS and Google.

We want to pay you for AI and Cloud services but we cannot because of this issue.

I realize consensus is needed at Google to support JPMS, so my sincere hope is that this genuine comment from an actual startup customer is motivating to the people who can decide to support JPMS. Feel free to screenshot or if you need me to email these comments to someone (Jeanine? We've spoken many times) I'm happy to do so

**Comment by esrauchg** [comment]
We will not moderate comments that simply disagree with us, as per none of your other comments being moderated. Please do use words rather than memes to make your arguments.

We hear and understand your feedback sgammon. I will pass your feedback on within Google including to the engineering team which maintains the Java Cloud Client Libraries and discuss this with other stakeholders internally. Passing your feedback on through your other available Cloud support channels is also a great idea.

If any one else has similar or different feedback here, please do let us know on this thread (and let your other cloud support channels know as well where appropriate).

Thanks

**Comment by jart** [comment]
Sam's a good guy and means well. It looks like Guava already moved on this. https://github.com/google/guava/commit/7a71ea0bfab8618765c2841111f998e484af6257 Everything you need to know is in that change. It'd be nice if protobuf could do this too.

**Comment by sgammon** [comment]
Thank you @esrauchg, I appreciate that greatly. I'll take your advice and pass this feedback through other channels we have access to as well

**Comment by cowwoc** [comment]
Please help me understand this point:

> I think concern here is the problem of additional constraints implied by supporting JMPS rather than the up front technical work of the declaration, including eg the implication of a module owning a package (by design you can't have multiple modules make visible the same package).

> That particular aspect be more of a hard blocking concern against adoption of JPMS for other Google published libraries like the Java Google Cloud Client Libraries than it is for Protobuf, but I believe this already breaks down even for how Protobuf currently packages our code for our Maven releases where com.google.protobuf package is used by several different artifacts.

If the `com.google.protobuf` package is used by several different artifacts, why not simply add new artifacts alongside the old ones? Specifically:

- [ ] Move classes into separate packages for each artifact.
- [ ] Publish JPMS-compatible artifacts for most users to consume
- [ ] Publish "backwards-compatibility" artifacts that use the old package names for users who are unable or unwilling to modify their imports (this can be a simple build-time task)

If you are really keen on minimizing changes for people who don't use JPMS, use the existing artifacts for the non-JPMS build and add new artifacts with JPMS support. Either way, many other libraries have done this and it works great!

**Comment by esrauchg** [comment]
Unfortunately there is no good way in Java to make the "backwards-compatibility artifacts" from your step three which can exist as to shim one package name on top of another one. Java lacks a feature that most other languages do have which can enable you to write "com.xyz.ByteString is an alias for com.xyz2.ByteString"; in the Java model two different fully-qualified-names must always be two different things.

So really what step 3 would really demand would be a total shading of the entire protobuf java runtime and publishing a protobuf-classic and a protobuf-jpms, with different contents, and that being 'viral' upwards to whoever uses us picking one of those two incompatible definitions and so on up the dependency stack. I don't think we would realistically go down that path for many reasons.

So the topic here really is whether Protobuf should fully take the commitment to the constraints which may restrict our ability to optimally serve the needs of the non-JPMS-users to better support JPMS-users. That is obviously a choice that is contextual for where the ecosystem is to understand that tradeoff (and yes, per sgammon's point, Protobuf is foundational enough that where we move will in turn have some influence where the ecosystem moves).

I will be following up so we can make an informed decision including trying to find out whether other major libraries are likely to adopt JPMS if Protobuf does, since if they don't also move I think this wouldn't actually help @sgammon, instead we would have effectively taken the constraints and not actually meaningfully unblocked the use in practice.

And I will stress: no one is going to rule out the eventual adoption of JPMS no matter what here, the only two possible options we may take here are "yes, today" and "not now, to be revisited as the situation evolves including e.g. maybe when more people finally move off of Java 8 they will want to use this shiny 'new' stuff"

**Comment by cowwoc** [comment]
@esrauchg I am proposing something much simpler than what you had in mind...

Once the source-code is refactored for JPMS compatibility, simply use a mix of https://maven.apache.org/plugins/maven-antrun-plugin/ and https://ant.apache.org/manual/Tasks/replaceregexp.html to rewrite the package names back to their original values at build time. Use the output to then build and publish "backwards-compatibility artifacts".

I've used this technique in the past. It works well, is fast, and is easy to maintain.

**Comment by esrauchg** [comment]
I might be misunderstanding but I think what you are describing is what I meant by shaded, and unfortunately I don't think we could ever realistically consider that kind of approach for Protobuf.

Imagine we have `x.CodedInputStream` in our current .jar and want to move it to a second jar. I believe your proposal is that we'd publish one jar with it as `x.CodedInputStream` (a back-compat artifact) and a jar with it as `x2.CodedInputStream` for the JPMS case.

Generated code names x.CodedInputStream, so any preexisting gencode which exists can't be used with that JPMS-compatible packaged one; then we'd also need a protoc flag that tells it whether to use the JPMS types or not and everyone using JPMS would need to turn on that flag to use the other package name. This would actually already be a deal-breaker for the plan.

But also then in in turn, any library which handles/ships gencode would itself need to pick if they use the JPMS-compat-gencode or the non-JPMS-compat-gencode or also ship two artifacts in turn themselves, and then anyone who uses _them_ would need to virally upward.

Under this this narrow example, we realistically would mitigate the opposite way, by eternally shipping one JPMS-compatible artifact which is the monolith merge of any other jars though.

**Comment by sgammon** [comment]
@esrauchg This thread is full of people who have fixed JPMS issues before. I think everyone here is aware of what JPMS enforces and why that might be problematic for Protocol Buffers.

We are requesting this feature regardless of those difficulties and if the Protobuf team needs help figuring it out, I'm quite sure we'd all jump in and come up with solutions for whatever stands in the way.

> ...find out whether other major libraries are likely to adopt JPMS if Protobuf does...

Protobuf is upstream from nearly everyone on earth. If Protobuf can't support JPMS, the world will eventually stop using Protobuf in Java or it will stop using JPMS. JPMS is part of Java and is used in the JDK, so it's not likely to be bumped out by an OSS library. In my view, this is a survival-level question for Google's Java libraries and should be treated with the highest priority, but I've already made my point there.

> I think this wouldn't actually help [me]

I appreciate you offering to talk to other library authors but really I just want Protobuf Java to work with JPMS, it has nothing to do with other code. There are other libraries blocking JPMS, but all of those rely on Protobuf, so Protobuf is where I am asking for that support.

Guava and other libraries were blocking Protobuf from adopting JPMS, but I believe that path is clear now.

> And I will stress: no one is going to rule out the eventual adoption of JPMS

Sure, yes. Nobody will say out loud, "we'll never support JPMS." It is just effectively what is going to happen anyway with no consensus. It is effectively the policy now.

**Comment by cowwoc** [comment]
> then we'd also need a protoc flag that tells it whether to use the JPMS types or not and everyone using JPMS would need to turn on that flag to use the other package name. This would actually already be a deal-breaker for the plan.

Yes, either add a flag or cut off non-JPMS support below a major version number. Which brings up another point: why is it a big deal asking users to regenerate their code when upgrading a major version number?

I always regenerate stubs at build-time, but for people who don't it doesn't sound like a big deal asking them to regenerate their code once.

My assumption is that we're asking users to regenerate the stubs and users of the stubs to upgrade the version of the protobuf library they depend on. Am I wrong? This seems like a very common ask in case of a major version upgrade...

Splitting those two concerns: regeneration of the stubs and updating the protobuf dependence, which one(s) is a showstopper for you, and why?

**Comment by esrauchg** [comment]
> I always regenerate stubs at build-time, but for people who don't it doesn't sound like a big deal asking them to regenerate their code once.

Yeah, we fundamentally can't.

This is a bit off-topic for the JPMS topic, but I'm hoping if I provide some "how the sausage is made" in response to that question it will help make clear why some obvious/easy solutions to problems are fundamentally not paths we could take, which in turn results in a cautious / very-slow-moving stance for topics like JPMS.

Many of our major customers unfortunately just do not have a good build system setup and they have very spaghetti dependencies. Identifying and regenerating all of their codegen can be extremely onerous. We had planned Java 4.x to require a "when you adopt 4.x you also need to update your codegen to 4.x", and we changed the plan because we found that would have resulted in 4.x just being rejected by some major customers.

Java protobuf 4.x today instead supports any codegen after 3.22 (where codegen from before that covered by a CVE so everyone should want to upgrade), and even the "incrementally get all of your codegen onto 3.25 first, then you can freely upgrade onto 4.x without updating codegen at that moment" can unfortunately be a herculean multi-year effort for some large customers.

Similarly we do unfortunately do have customers who refuse new versions of Py Proto today due to a codegen incompatibility point a number of years back from today (see https://github.com/protocolbuffers/protobuf/issues/10051).

**Comment by sgammon** [comment]
Maybe this could be solved with an additional plugin? Split packages and gencode seem like two problems, really. I'm interested to hear more once your response is ready @esrauchg. JPMS is a puzzle which often needs some escape hatch or trick somewhere for proper backwards compat. On Guava and other libs, we've needed a lot of these to check upstream boxes.

Once you can share, please do, and I know @cowwoc and I would be happy to help think of ways to turn this into a win for everyone. Thanks also for being responsive on this issue and for doing the legwork/research internally.

**Comment by odinsbane** [comment]
Named modules cannot share a split package, and that seems to be the case going forward with java. According to this SO answer, https://stackoverflow.com/a/51828900/2067492 if two libraries share a split package than one of them can be an explicit named module provided the other would remain in the unnamed module. 

Maybe protobuf should become a named module first? That shouldn't break anything because the other libraries that share a split package would remain in the unnamed module. Which would remain backwards compatible. I don't know how to test that though.

I haven't found anything to suggest that java will eventually provide a way to use modules with split packages.

**Comment by cowwoc** [comment]
@esrauchg Would you accept this alternative?

* Continue publishing the current (classpath) artifact as-is.
* Publish a new modular artifact that uses a new package name and generates modular classes.
* Have the classpath artifact's classes (and its generated classes) extend the modular artifact's classes.
* (Optional but recommended) Mark the classpath classes as `@Deprecated` in favor of the modular classes.

Let me know if this approach checks all the necessary boxes. I'd like to begin working on a PR.

**Comment by esrauchg** [comment]
Sorry for the slow updates here. The tentative situation here is still that our other major users on top (like Cloud SDK) don't currently have intent to adopt JPMS, and Guava did have more pain points than they had hoped when they accepting external patches for this topic.

So the wheels are turn slowly, but as of Q3-2025 I think we're not curerntly willing to go down the path of trying to publish the JPMS package based on the current cost-benefit, where the implementation of the up-front patches aren't a big part of this equation.

> @esrauchg Would you accept this alternative?

Unfortunately I don't think this path works out due to inheritance hierarchies. If you make a new `protobuf2.GeneratedMessage extends protobuf.GeneratedMessage` and then a `protobuf2.AbstractMessage extends protobuf.AbstractMessage`,  you won't be able to have `protobuf2.GeneratedMessage extends protobuf2.AbsrtactMessage`.

If such an approach could work without that gotcha, I think a path to get progress here would be to publish the package to maven of a wrapping package (`com.cowoc.Protobuf`) with a permissive license that anyone can use if they have a strong JPMS need, and then it could be considered for upstreaming from there.

Thanks all and sorry for the slow progress on this topic.
